### PR TITLE
Zero shot

### PR DESCRIPTION
**Pull request by Oufattole**
Resolves issue #105 .

- [x] Added a time to event labeler where a user can select a start and end time window (after the prediction time), and check for presence of a list of codes, or whether any of these codes are within a numerical value interval.
- [x] Make this labeler selectable via hydra so users can override with their own
- [x] Added a to_trajectory_batch function that prepares raw trajectories to a meds-like tensor that the labeler can reason over.

### CODE REVIEW

File: src/meds_torch/configs/model/backbone/eic_transformer_decoder.yaml [hunk 1]
```diff
@@ -1,4 +1,5 @@
 defaults:
+  - generation_budget: seq_len
```
Comment (mmcdermott) [inline src/meds_torch/configs/model/backbone/eic_transformer_decoder.yaml:2]: I wonder if there is a way to do this that doesn't inject it into the generic decoder config. I'm not sure it is a problem if it is here, but it is worth a (brief) thought.

File: src/meds_torch/configs/model/zero_shot_labeler/time_to_event_labeler.yaml
Comment (mmcdermott) [inline src/meds_torch/configs/model/zero_shot_labeler/time_to_event_labeler.yaml:1]: It is worth thinking about whether or not there can be a more general labeler template, potentially based on ACES configuration files.

File: src/meds_torch/configs/predict.yaml [hunk 2]
```diff
@@ -9,6 +9,20 @@ defaults:
   - paths: default
   - extras: default
   - hydra: default
+  # experiment configs allow for version control of specific hyperparameters
+  # e.g. best hyperparameters for given model and datamodule
+  - experiment: null
+
+  # config for hyperparameter optimization
+  - hparams_search: null
```
Comment (mmcdermott) [inline src/meds_torch/configs/predict.yaml:17]: Why do you need this in a prediction config?

File: src/meds_torch/utils/custom_normalization.py [hunk 3]
```diff
@@ -1,6 +1,9 @@
 #!/usr/bin/env python
 """Transformations for normalizing MEDS datasets, across both categorical and continuous
-dimensions."""
+dimensions.
+
+DO NOT RUN THIS WITH PARALLELISM
```
Comment (mmcdermott) [inline src/meds_torch/utils/custom_normalization.py:5]: Might be worth putting this in the main function docstring; then it will show up on help strings too I think.

File: src/meds_torch/utils/utils.py [hunk 4]
```diff
@@ -69,7 +71,7 @@ def wrap(cfg: DictConfig, **kwargs) -> tuple[dict[str, Any], dict[str, Any]]:
         # things to do if exception occurs
         except Exception as ex:
```
Comment (mmcdermott) [inline src/meds_torch/utils/utils.py:72]: Could add a `pragma: no cover` comment here to avoid the code cov error given this is just logging.

File: src/meds_torch/models/components/transformer_decoder.py [hunk 5]
```diff
@@ -1,15 +1,286 @@
...
+
+        >>> # Direct initialization isn't intended for use
+        >>> budget = GenerationBudget(100)
+    """
+
+    def __init__(self, value: int):
+        self._value = value
+
+    @classmethod
+    def from_seq_len(cls, value: int) -> "GenerationBudget":
```
Comment (mmcdermott) [inline src/meds_torch/models/components/transformer_decoder.py:70]: Why not just have two different init params -- one for seq len and one for time? that seems like it would make the configs clearer and simplify this code.

File: src/meds_torch/models/components/transformer_decoder.py [hunk 6]
```diff
@@ -1,15 +1,286 @@
...
+        Returns:
+            A GenerationBudget instance with max_seq_len set
+        """
+        instance = cls(value)
+        instance._is_seq_len = True
+        return instance
+
+    @classmethod
+    def from_time_len(cls, value: int) -> "GenerationBudget":
+        """Create a GenerationBudget from a minimum time length.
```
Comment (mmcdermott) [inline src/meds_torch/models/components/transformer_decoder.py:85]: For functions this simple a one-line docstring is acceptable. https://google.github.io/styleguide/pyguide.html#383-functions-and-methods

File: src/meds_torch/models/components/transformer_decoder.py [hunk 7]
```diff
@@ -1,15 +1,286 @@
...
+        instance = cls(value)
+        return instance
+
+    @property
+    def max_seq_len(self) -> int | None:
+        """The maximum sequence length, if this budget was created with from_seq_len."""
+        return self._value if hasattr(self, "_is_seq_len") else None
+
+    @property
+    def min_time_len(self) -> int | None:
```
Comment (mmcdermott) [inline src/meds_torch/models/components/transformer_decoder.py:102]: I don't know if this is sufficiently complex -- how do you handle "leave dataset" events? E.g., if a patient dies, or at the end of their covered data?

File: src/meds_torch/models/components/transformer_decoder.py [hunk 8]
```diff
@@ -1,15 +1,286 @@
...
+        >>> continue_gen # Should stop since > max_seq_len
+        False
+    """
+    # Get budget constraints
+    min_time_len = budget.min_time_len
+    is_time_budget = exists(min_time_len)
+    if is_time_budget:
+        # Get time prediction for the new token
+        with torch.no_grad():
+            if get_next_token_time is None:
```
Comment (mmcdermott) [inline src/meds_torch/models/components/transformer_decoder.py:193]: I would pull this error case out into the top to simplify your flow -- e.g., `if is_time_budget and get_next_token_time is None: ...` and also test it in your doctests.

File: src/meds_torch/models/components/transformer_decoder.py [hunk 9]
```diff
@@ -1,15 +1,286 @@
...
+
+
 class TransformerDecoderModel(torch.nn.Module, Module):
-    """Wrapper of Decoder Transformer for use in MEDS with triplet token embeddings."""
+    """Wrapper of Decoder Transformer for use in MEDS with triplet token embeddings.
+
+    This model handles both forward passes and generation with different budget types.
+
+    Examples:
+        >>> # Setup mock configuration
```
Comment (mmcdermott) [inline src/meds_torch/models/components/transformer_decoder.py:211]: I'd put some/all of this setup in a function so that these doctests are easier to read.

File: src/meds_torch/models/components/transformer_decoder.py [hunk 10]
```diff
@@ -32,3 +298,286 @@ def forward(self, batch):
...
+        eos_token=None,
+        temperature=1.0,
+        filter_logits_fn: str | Callable = identity,
+        restrict_to_max_seq_len: bool = True,
+        filter_kwargs: dict = dict(),
+        cache_kv=True,
+        pad_value: int = 0,
+        **kwargs,
+    ) -> torch.Tensor:
+        """Generate tokens with either sequence length or time budget constraints.
```
Comment (mmcdermott) [inline src/meds_torch/models/components/transformer_decoder.py:316]: What's the point of this function rather than just using the base `generate`? That might be worth commenting on in this docstring.

File: src/meds_torch/models/components/transformer_decoder.py [hunk 11]
```diff
@@ -32,3 +298,286 @@ def forward(self, batch):
...
+    >>> budget = GenerationBudget.from_seq_len(2)
+    >>> prompts = torch.randint(0, vocab_size, (B, S))
+    >>> gen_tokens = generate(model=model,
+    ...     prompts=prompts,
+    ...     budget=budget,
+    ...     temperature=1.0,
+    ...     filter_logits_fn='top_k',
+    ...     filter_kwargs={'k': 10},
+    ... )
+    >>> gen_tokens.shape  # Should be [batch_size, seq_len]
```
Comment (mmcdermott) [inline src/meds_torch/models/components/transformer_decoder.py:413]: These tests are pretty limited, and I don't know if you really need all of them; you could also simplify them by sharing parameters across cases directly to make it more useful as documentation.

File: src/meds_torch/models/components/transformer_decoder.py [hunk 12]
```diff
@@ -32,3 +298,286 @@ def forward(self, batch):
...
+            filter_kwargs=filter_kwargs,
+            cache_kv=cache_kv,
+            pad_value=pad_value,
+            **kwargs,
+        )
+
+
+@torch.no_grad()
+@eval_decorator
+def generate(
```
Comment (mmcdermott) [inline src/meds_torch/models/components/transformer_decoder.py:358]: This stuff seems very specific to the EIC case. should this be in a more specific file?

File: src/meds_torch/models/components/transformer_decoder.py
Comment (mmcdermott) [inline src/meds_torch/models/components/transformer_decoder.py:1]: Are there integration tests for this, and if so where?

File: src/meds_torch/predict.py [hunk 13]
```diff
@@ -188,6 +219,51 @@ def predict(cfg: DictConfig, datamodule=None) -> tuple[dict[str, Any], dict[str,
         cfg: DictConfig configuration composed by Hydra.
     Returns:
         Tuple[dict, dict] with metrics and dict with all instantiated objects.
+
+    Examples:
+    >>> import tempfile
+    >>> from omegaconf import DictConfig
+    >>> _ = pl.Config.set_tbl_width_chars(106)
+    >>> # Create temporary checkpoint file
```
Comment (mmcdermott) [inline src/meds_torch/predict.py:227]: you should mention in the docstring that you're using the dummy classes you created above.

File: src/meds_torch/utils/pylogger.py
Comment (mmcdermott) [inline src/meds_torch/utils/pylogger.py:1]: This seems like a generally useful pytorch logging utility. Is there not something like this as a separate package? If not, you should consider spending (time box this!) an hour pulling this out into a separate package so that it could be easily used in other projects.

File: src/meds_torch/models/zero_shot_labeler/time_to_event_labeler.py [hunk 14]
```diff
@@ -0,0 +1,140 @@
+from dataclasses import dataclass
+
+import torch
+
+from meds_torch.models.zero_shot_labeler.utils import TrajectoryBatch
+
+
+@dataclass
+class TaskLabeler:
```
Comment (mmcdermott) [inline src/meds_torch/models/zero_shot_labeler/time_to_event_labeler.py:9]: Is this a general class or for a specific kind of task only? If the latter, I suggest a more specific name.

File: src/meds_torch/models/eic_forecasting.py [hunk 15]
```diff
@@ -19,10 +20,71 @@
...
+    def generate(
+        self,
+        prompts: torch.Tensor,
+        mask: torch.Tensor | None = None,
+        get_next_token_time: Callable[[torch.Tensor, torch.Tensor], torch.Tensor] | None = None,
+        temperature: float = 1.0,
+        filter_logits_fn: str | Callable = torch.nn.Identity(),
+    ) -> torch.Tensor:
+        B, S = prompts.shape
+        out = torch.randint(0, 4, (B, S))
```
Comment (mmcdermott) [inline src/meds_torch/models/eic_forecasting.py:51]: I'd have this and below return deterministic data, just for ease of testing.

File: src/meds_torch/models/eic_forecasting.py [hunk 16]
```diff
@@ -117,9 +179,10 @@ def __init__(self, vocab_size: int, dist_sync_on_step=False):
         self.vocab_size = vocab_size
 
         self.auroc = MulticlassAUROC(num_classes=vocab_size, average="weighted", thresholds=100)
-        self.top_1_accuracy = MulticlassAccuracy(num_classes=vocab_size, top_k=1)
-        self.top_5_accuracy = MulticlassAccuracy(num_classes=vocab_size, top_k=5)
-        self.top_10_accuracy = MulticlassAccuracy(num_classes=vocab_size, top_k=10)
+        self.top_k_acc = top_k_acc
+        self.accuracy_metrics = MetricCollection(
```
Comment (mmcdermott) [inline src/meds_torch/models/eic_forecasting.py:183]: Why not put AUROC into the metric collection too? Then just have all metrics in one collection?

File: src/meds_torch/models/eic_forecasting.py [hunk 17]
```diff
@@ -162,31 +223,163 @@ def compute(self):
         results = {
             "auroc": self.auroc.compute(),
         }
-        results["top_1_accuracy"] = self.top_1_accuracy.compute()
-        results["top_5_accuracy"] = self.top_5_accuracy.compute()
-        results["top_10_accuracy"] = self.top_10_accuracy.compute()
+        results.update(self.accuracy_metrics.compute())
```
Comment (mmcdermott) [inline src/meds_torch/models/eic_forecasting.py:226]: Just put this in the main dictionary like AUROC too -- no reason to use two dictionary syntaxes.

File: src/meds_torch/models/eic_forecasting.py [hunk 18]
```diff
@@ -162,31 +223,163 @@ def compute(self):
...
+    >>>  # check that 1 time token was generated for the shortest time trajectory
+    >>> (codes == 2).sum(dim=1).min().item()
+    1
+    >>> temp_file.close()
+    """
 
     def __init__(self, cfg: DictConfig):
         super().__init__(cfg)
-        if not isinstance(self.model, AUTOREGRESSIVE_MODELS):
+        if not isinstance(self.model, AUTOREGRESSIVE_MODELS + (DummyModel,)):
```
Comment (mmcdermott) [inline src/meds_torch/models/eic_forecasting.py:369]: You don't want to have to modify your code for these sorts of testing settings... Is there any way to put `DummyModel` in the list of `AUTOREGRESSIVE_MODELS` somehow, either in general if appropriate or only during doctests? Similarly for the encoder

File: src/meds_torch/utils/custom_filter_measurements.py
Comment (mmcdermott) [inline src/meds_torch/utils/custom_filter_measurements.py:1]: This can eventually be added to MEDS-Transforms directly

**Review by mmcdermott — APPROVED** [review]
I left a number of comments but don't feel blocked by my review; proceed at your own target pace.

File: src/meds_torch/configs/model/zero_shot_labeler/time_to_event_labeler.yaml
Comment (Oufattole) [reply src/meds_torch/configs/model/zero_shot_labeler/time_to_event_labeler.yaml:1]: That's in the works here: https://github.com/Oufattole/clinical-zeroshot-labeler

File: src/meds_torch/configs/model/backbone/eic_transformer_decoder.yaml (hunk 1 above)
Comment (Oufattole) [reply src/meds_torch/configs/model/backbone/eic_transformer_decoder.yaml:2]: I think I can move this (in a later commit to be at the eic model level. as I can add a _target_ in the eic model yaml which will call the czsl (clinical zero-shot labeler) library, which handles stop criterias and creating labels.

File: src/meds_torch/configs/predict.yaml (hunk 2 above)
Comment (Oufattole) [reply src/meds_torch/configs/predict.yaml:17]: Removed thanks!

File: src/meds_torch/models/components/transformer_decoder.py (hunk 5 above)
Comment (Oufattole) [reply src/meds_torch/models/components/transformer_decoder.py:70]: Great Idea, I'll do that in the czsl library.
